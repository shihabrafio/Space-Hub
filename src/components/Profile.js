import React from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { cancelReservation } from '../redux/dragons/dragonsSlice';
import { cancelRocket } from '../redux/rockets/rocketsSlice';

const Profile = () => {
  const { dragons } = useSelector((state) => state.dragons);
  const reservedDragons = dragons.filter((dragon) => dragon.reserved);

  const { rockets } = useSelector((state) => state.rockets);
  const reservedRockets = rockets.filter((rocket) => rocket.reserved);

  const missions = useSelector((state) => state.missions.missions);
  const filterMissions = missions.filter((mission) => mission.status === true);

  const dispatch = useDispatch();

  return (
    <section className="profile-container">
      <div className="dragons-wrapper">
        <h2>My Dragons</h2>
        <table>
          <tbody>
            {reservedDragons.length ? (
              reservedDragons.map((dragon) => (
                <tr className="profile-tr" key={dragon.id}>
                  <td>{dragon.name}</td>
                  <button
                    className="btn secondary-outline-btn"
                    onClick={() => dispatch(cancelReservation(dragon.id))}
                    type="button"
                  >
                    Cancel Reservation
                  </button>
                </tr>
              ))
            ) : (
              <tr>
                <td>No Dragons reserved!</td>
              </tr>
            )}
          </tbody>
        </table>
      </div>
      <div className="rockets-wrapper">
        <h2>My Rockets</h2>
        <table>
          <tbody>
            {reservedRockets.length ? (
              reservedRockets.map((rocket) => (
                <tr className="profile-tr" key={rocket.id}>
                  <td>{rocket.name}</td>
                  <button
                    className="btn secondary-outline-btn"
                    onClick={() => dispatch(cancelRocket(rocket.id))}
                    type="button"
                  >
                    Cancel Reservation
                  </button>
                </tr>
              ))
            ) : (
              <tr>
                <td>No Rockets reserved!</td>
              </tr>
            )}
          </tbody>
        </table>
      </div>
      <div className="missions-wrapper">
        <h2>My Mission</h2>
        <table>
          <tbody>
            {filterMissions.length ? (
              filterMissions.map((mission) => (
                <tr key={mission.id}>
                  <td>{mission.name}</td>
                </tr>
              ))
            ) : (
              <tr>
                <td>No Missions Joined!</td>
              </tr>
            )}
          </tbody>
        </table>
      </div>
    </section>
  );
};

export default Profile;
