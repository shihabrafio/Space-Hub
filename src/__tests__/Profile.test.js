import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import Profile from '../components/Profile';
import dragonsReducer from '../redux/dragons/dragonsSlice';
import rocketsReducer from '../redux/rockets/rocketsSlice';
import missionsReducer from '../redux/missions/missionsSlice';

const store = configureStore({
  reducer: {
    dragons: dragonsReducer,
    rockets: rocketsReducer,
    missions: missionsReducer,
  },
});

describe('Profile component', () => {
  test('renders with reserved dragons', () => {
    const reservedDragons = [
      { id: 1, name: 'Dragon 1', reserved: true },
      { id: 2, name: 'Dragon 2', reserved: true },
    ];

    const { container } = render(
      <Provider store={store}>
        <Profile
          reservedDragons={reservedDragons}
          reservedRockets={[]}
          activeMissions={[]}
        />
      </Provider>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders with reserved rockets', () => {
    const reservedRockets = [
      { id: 1, name: 'Rocket 1', reserved: true },
      { id: 2, name: 'Rocket 2', reserved: true },
    ];

    const { container } = render(
      <Provider store={store}>
        <Profile
          reservedDragons={[]}
          reservedRockets={reservedRockets}
          activeMissions={[]}
        />
      </Provider>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders with active missions', () => {
    const activeMissions = [
      { id: 1, name: 'Mission 1', status: true },
      { id: 2, name: 'Mission 2', status: true },
    ];

    const { container } = render(
      <Provider store={store}>
        <Profile
          reservedDragons={[]}
          reservedRockets={[]}
          activeMissions={activeMissions}
        />
      </Provider>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders with no reserved dragons or rockets', () => {
    const { container } = render(
      <Provider store={store}>
        <Profile
          reservedDragons={[]}
          reservedRockets={[]}
          activeMissions={[]}
        />
      </Provider>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders with no active missions', () => {
    const { container } = render(
      <Provider store={store}>
        <Profile
          reservedDragons={[]}
          reservedRockets={[]}
          activeMissions={[]}
        />
      </Provider>
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
