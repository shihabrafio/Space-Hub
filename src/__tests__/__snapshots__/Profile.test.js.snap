// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile component renders with active missions 1`] = `
<section
  class="profile-container"
>
  <div
    class="dragons-wrapper"
  >
    <h2>
      My Dragons
    </h2>
    <table>
      <tbody>
        <tr>
          <td>
            No Dragons reserved!
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    class="rockets-wrapper"
  >
    <h2>
      My Rockets
    </h2>
    <table>
      <tbody>
        <tr>
          <td>
            No Rockets reserved!
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    class="missions-wrapper"
  >
    <h2>
      My Mission
    </h2>
    <table>
      <tbody>
        <tr>
          <td>
            No Missions Joined!
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</section>
`;

exports[`Profile component renders with no active missions 1`] = `
<section
  class="profile-container"
>
  <div
    class="dragons-wrapper"
  >
    <h2>
      My Dragons
    </h2>
    <table>
      <tbody>
        <tr>
          <td>
            No Dragons reserved!
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    class="rockets-wrapper"
  >
    <h2>
      My Rockets
    </h2>
    <table>
      <tbody>
        <tr>
          <td>
            No Rockets reserved!
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    class="missions-wrapper"
  >
    <h2>
      My Mission
    </h2>
    <table>
      <tbody>
        <tr>
          <td>
            No Missions Joined!
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</section>
`;

exports[`Profile component renders with no reserved dragons or rockets 1`] = `
<section
  class="profile-container"
>
  <div
    class="dragons-wrapper"
  >
    <h2>
      My Dragons
    </h2>
    <table>
      <tbody>
        <tr>
          <td>
            No Dragons reserved!
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    class="rockets-wrapper"
  >
    <h2>
      My Rockets
    </h2>
    <table>
      <tbody>
        <tr>
          <td>
            No Rockets reserved!
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    class="missions-wrapper"
  >
    <h2>
      My Mission
    </h2>
    <table>
      <tbody>
        <tr>
          <td>
            No Missions Joined!
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</section>
`;

exports[`Profile component renders with reserved dragons 1`] = `
<section
  class="profile-container"
>
  <div
    class="dragons-wrapper"
  >
    <h2>
      My Dragons
    </h2>
    <table>
      <tbody>
        <tr>
          <td>
            No Dragons reserved!
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    class="rockets-wrapper"
  >
    <h2>
      My Rockets
    </h2>
    <table>
      <tbody>
        <tr>
          <td>
            No Rockets reserved!
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    class="missions-wrapper"
  >
    <h2>
      My Mission
    </h2>
    <table>
      <tbody>
        <tr>
          <td>
            No Missions Joined!
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</section>
`;

exports[`Profile component renders with reserved rockets 1`] = `
<section
  class="profile-container"
>
  <div
    class="dragons-wrapper"
  >
    <h2>
      My Dragons
    </h2>
    <table>
      <tbody>
        <tr>
          <td>
            No Dragons reserved!
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    class="rockets-wrapper"
  >
    <h2>
      My Rockets
    </h2>
    <table>
      <tbody>
        <tr>
          <td>
            No Rockets reserved!
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    class="missions-wrapper"
  >
    <h2>
      My Mission
    </h2>
    <table>
      <tbody>
        <tr>
          <td>
            No Missions Joined!
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</section>
`;
